### PR TITLE
Replace test doubles with constants

### LINKED ISSUE #2321: Replace all weird numbers with constant values
### What should be done?

Currently in integration tests we have a couple of values that look like this:
```
"double-7":                  1.79769e+307,
"double-max-overflow":       9.223372036854776833e+18,
"double-max-overflow-verge": 9.223372036854776832e+18,
```

It's difficult to understand what exactly they represent. 
For example, because of that reason we use `math.MaxInt64` in place of `9223372036854775807`, and `doubleMaxPrec` (which is defined with bitshift operator `1<<53 - 1`) in place of `9007199254740991.0`.

We should identify what exactly those values represent and replace them with calls to constants.

If it's possible - let's simplify creation of those constants by using bitshift operators or simple arithmetic, so they are "human-readable".

There's no need to replace values like this though, they look pretty straightforward (it's just a random larger integer, parsed to float64):
```
"double-4":                  float64(123456789)
```

### Where?

- shareddata package

### Definition of Done

- find all unclear numbers and identify them;
- create constants for those numbers, add comments for them;
- replace all numbers with created and/or existing constants;
- remove all TODO references to this issue.

### PR DESCRIPTION
**Pull request by noisersup**
# Description

Closes #2321.

## Readiness checklist

- [ ] I added/updated unit tests.
- [x] I added/updated integration/compatibility tests.
- [ ] I added/updated comments and checked rendering.
- [x] I made spot refactorings.
- [ ] I updated user documentation.
- [ ] I ran `task all`, and it passed.
- [x] I ensured that PR title is good enough for the changelog.
- [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Labels, Project and project's Sprint fields.
- [x] I marked all done items in this checklist.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -250,6 +250,14 @@ In most cases, they should be used instead of (deprecated) `bson.D.Map()`,
 The bar for adding new helpers is very high.
 Please check all existing ones.
 
+If there's a need to use any large number to test corner cases, 
+we create constants for them with explanation what do they represent, and refer to them.
+For example:
+
+```
+const doubleMaxPrec = float64(1<<53 - 1) // 9007199254740991.0:    largest double values that could be represented as integer exactly
```
Comment (chilagrow) [inline CONTRIBUTING.md:258]: In our code base we have mixture of `float64(9007199254740991)` and `9007199254740991.0`. More commonly found is `0.0` and `float64(0)`. @quasilyte identified this as something we can make more consistent via linter. What is our preference? This documentation just reminded me of that :hugs:

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -250,6 +250,14 @@ In most cases, they should be used instead of (deprecated) `bson.D.Map()`,
 The bar for adding new helpers is very high.
 Please check all existing ones.
 
+If there's a need to use any large number to test corner cases, 
+we create constants for them with explanation what do they represent, and refer to them.
+For example:
+
+```
```
Comment (chilagrow) [inline CONTRIBUTING.md:257]: Should it be ```go
?

File: integration/shareddata/scalars.go [hunk 3]
```diff
@@ -28,7 +28,10 @@ const (
 	doubleBig     = float64(1 << 61)   // 2305843009213693952.0: some number larger than safe integer (doubleBig+1 == doubleBig)
 	longBig       = int64(1 << 61)     // 2305843009213693952:   same as doubleBig but integer
 
-	// TODO https://github.com/FerretDB/FerretDB/issues/2321
+	doubleMaxOverflow = float64(9223372036854776833)  // double value that is larger than MaxInt64
+	doubleMinOverflow = float64(-9223372036854776833) // double value that is smaller than MinInt64
+
+	doubleMaxVerge = 1.79769e+307 // double value that is close to the MaxFloat64 (1.79769e+308)
```
Comment (chilagrow) [inline integration/shareddata/scalars.go:34]: Could you also add non scientific notation in a comment?

What is the special thing about constant `1.79769`? I know it's the same constant as `MaxFloat64`. But does this mean that adding certain number will make it overflow or?

File: integration/shareddata/scalars.go [hunk 4]
```diff
@@ -28,7 +28,10 @@ const (
 	doubleBig     = float64(1 << 61)   // 2305843009213693952.0: some number larger than safe integer (doubleBig+1 == doubleBig)
 	longBig       = int64(1 << 61)     // 2305843009213693952:   same as doubleBig but integer
 
-	// TODO https://github.com/FerretDB/FerretDB/issues/2321
+	doubleMaxOverflow = float64(9223372036854776833)  // double value that is larger than MaxInt64
```
Comment (chilagrow) [inline integration/shareddata/scalars.go:31]: Could you also check these int64? I think I moved them to shareddata when we move from integration to compat tests. But I don't know the significance of those numbers :thinking: 
```
		"int64-1":          int64(1099511628000),
		"int64-2":          int64(281474976700000),
		"int64-3":          int64(72057594040000000),
```
,
```
		"int32-1": int32(4080),
		"int32-2": int32(1048560),
		"int32-3": int32(268435440),
```

and
```
	"double-1":     4080.1234,
		"double-2":     1048560.0099,
		"double-3":     268435440.2,
```

**Review by chilagrow — COMMENTED** [review]
Great work so far! Few comments :hugs:

File: integration/shareddata/scalars.go (hunk 4 above)
Comment (chilagrow) [reply integration/shareddata/scalars.go:31]: Looking at PR title, maybe there will be another PR for handling non-double numbers?

File: integration/shareddata/scalars.go (hunk 3 above)
Comment (noisersup) [reply integration/shareddata/scalars.go:34]: Do we want to represent `1.79769e+307` in standard notation? It will be huge comment...

> What is the special thing about constant 1.79769? I know it's the same constant as MaxFloat64.

It's not, this one is on a verge (`+307`) and `MaxFloat64` has (`+308`)

File: integration/shareddata/scalars.go (hunk 4 above)
Comment (noisersup) [reply integration/shareddata/scalars.go:31]: I\ve checked all of those values and it seems to me that none of them represent any specific number.

File: CONTRIBUTING.md (hunk 1 above)
Comment (noisersup) [reply CONTRIBUTING.md:258]: I guess in simpler cases like `0.0` it seems better to use this notation

File: integration/shareddata/scalars.go (hunk 4 above)
Comment (AlekSi) [reply integration/shareddata/scalars.go:31]: So let's add a comment about that. `// random` or something like that

File: integration/shareddata/scalars.go (hunk 3 above)
Comment (AlekSi) [reply integration/shareddata/scalars.go:34]: The difference between 1.79769e+308 and 1.79769e+307 is 1.617921e+308; I would not call those values "close".

Could it be `math.MaxFloat64 - 1.0`?

File: integration/shareddata/scalars.go [hunk 5]
```diff
@@ -28,7 +28,10 @@ const (
...
+	doubleMinOverflow = float64(-9223372036854776833) // double value that is smaller than MinInt64
```
Comment (AlekSi) [inline integration/shareddata/scalars.go:32]: So could it be `float64(math.MinInt64) - 1.0)`, with a value in a comment?

Overall, we have a strange mix of expressions and literals.

Comment (noisersup) [reply integration/shareddata/scalars.go:32]: subtrackting 1.0 will probably result in the same number because of the lack of precision

Comment (AlekSi) [reply integration/shareddata/scalars.go:32]: -2.0 then? After all, we have a test for that in init

File: integration/shareddata/scalars.go (hunk 3 above)
Comment (chilagrow) [reply integration/shareddata/scalars.go:34]: If tests work with `math.MaxFloat64 - 1.0`, calling that `verge` makes more sense to me :hugs:

File: CONTRIBUTING.md (hunk 1 above)
Comment (chilagrow) [reply CONTRIBUTING.md:258]: Okie :+1:

**Comment by chilagrow** [comment]
Tests failing for `TestQueryBitwise*`

File: integration/shareddata/scalars.go [hunk 6]
```diff
@@ -147,17 +148,15 @@ var Doubles = &Values[string]{
...
 
-		"double-null":               nil,
-		"double-1":                  float64(math.MinInt64 - 1),
-		"double-2":                  float64(math.MinInt64),
-		"double-3":                  float64(-123456789),
-		"double-4":                  float64(123456789),
-		"double-5":                  float64(math.MaxInt64),
-		"double-6":                  float64(math.MaxInt64 + 1),
-		"double-max-overflow":       9.223372036854776833e+18,
-		"double-max-overflow-verge": 9.223372036854776832e+18,
```
Comment (chilagrow) [inline integration/shareddata/scalars.go:158]: I thought removing this would decrease test coverage of `$inc` or `$mul` query, or maybe aggregation `$sum` :thinking:

**Review by rumyantseva — APPROVED** [review]
LGTM

**Review by chilagrow — DISMISSED** [review]

File: integration/shareddata/scalars.go [hunk 7]
```diff
@@ -24,36 +24,37 @@ import (
 )
 
 const (
-	doubleMaxPrec = float64(1<<53 - 1) // 9007199254740991.0:    largest double values that could be represented as integer exactly
-	doubleBig     = float64(1 << 61)   // 2305843009213693952.0: some number larger than safe integer (doubleBig+1 == doubleBig)
-	longBig       = int64(1 << 61)     // 2305843009213693952:   same as doubleBig but integer
+	doubleMaxPrec float64 = float64(1<<53 - 1) // 9007199254740991.0:    largest double values that could be represented as integer exactly
```
Comment (AlekSi) [inline integration/shareddata/scalars.go:27]: There is no need to specify type twice – it looks intentional, but actually does nothing and is confusing

File: integration/shareddata/scalars.go [hunk 8]
```diff
@@ -24,36 +24,37 @@ import (
...
+	doubleBig     float64 = float64(1 << 61)   // 2305843009213693952.0: some number larger than safe integer (doubleBig+1 == doubleBig)
+	longBig       int64   = int64(1 << 61)     // 2305843009213693952:   same as doubleBig but integer
```
Comment (AlekSi) [inline integration/shareddata/scalars.go:29]: Can it be `int64(doubleBig)` then if it is the same? And then we probably test in the `init` function that it is actually the same

File: integration/shareddata/scalars.go [hunk 9]
```diff
@@ -24,36 +24,37 @@ import (
...
-		"double-max-overflow-verge": 9.223372036854776832e+18,
-		"double-min-overflow":       -9.223372036854776833e+18,
-		"double-min-overflow-verge": -9.223372036854776832e+18,
+		"double":              42.13,
+		"double-whole":        42.0,
+		"double-zero":         0.0,
+		"double-max":          math.MaxFloat64,
+		"double-smallest":     math.SmallestNonzeroFloat64,
+		"double-big":          doubleBig,
+		"double-1":            float64(math.MinInt64 - 1),
```
Comment (AlekSi) [inline integration/shareddata/scalars.go:49]: But we already have a constant for "double value that is smaller than MinInt64" (`doubleMinOverflow`). Why not use it?

File: integration/shareddata/scalars.go [hunk 10]
```diff
@@ -24,36 +24,37 @@ import (
...
+		"double-2":            float64(math.MinInt64),
+		"double-3":            float64(-123456789), // random
+		"double-4":            float64(123456789),  // random
+		"double-5":            float64(math.MaxInt64),
+		"double-6":            float64(math.MaxInt64 + 1),
```
Comment (AlekSi) [inline integration/shareddata/scalars.go:54]: `doubleMaxOverflow`?

File: integration/shareddata/scalars.go [hunk 11]
```diff
@@ -102,10 +103,10 @@ var Scalars = &Values[string]{
 		"int64-max":        int64(math.MaxInt64),
 		"int64-min":        int64(math.MinInt64),
 		"int64-big":        longBig,
-		"int64-double-big": int64(1 << 61),
-		"int64-1":          int64(1099511628000),
-		"int64-2":          int64(281474976700000),
-		"int64-3":          int64(72057594040000000),
+		"int64-double-big": doubleBig,
+		"int64-1":          int64(1099511628000),     // random,
```
Comment (AlekSi) [inline integration/shareddata/scalars.go:107]:
```suggestion
		"int64-1":          int64(1099511628000),     // random
```

File: integration/shareddata/scalars.go [hunk 12]
```diff
@@ -353,4 +348,9 @@ func init() {
 
 	must.BeTrue(-doubleMaxPrec != -doubleMaxPrec-1)
 	must.BeTrue(-doubleMaxPrec-1 == -doubleMaxPrec-2)
+
+	must.BeTrue(doubleMaxOverflow > float64(math.MaxInt64))
+	must.BeTrue(doubleMinOverflow <= float64(math.MinInt64))
```
Comment (AlekSi) [inline integration/shareddata/scalars.go:353]:
```suggestion
	must.BeTrue(doubleMinOverflow < float64(math.MinInt64))
```

Otherwise, we don't check much

File: integration/shareddata/scalars.go (hunk 8 above)
Comment (AlekSi) [reply integration/shareddata/scalars.go]: The second part wasn't done

File: integration/shareddata/scalars.go [hunk 13]
```diff
@@ -26,34 +26,33 @@ import (
...
+		"double-whole":        42.0,
+		"double-zero":         0.0,
+		"double-max":          math.MaxFloat64,
+		"double-smallest":     math.SmallestNonzeroFloat64,
+		"double-big":          doubleBig,
+		"double-1":            float64(math.MinInt64),
+		"double-2":            float64(-123456789), // random
+		"double-3":            float64(123456789),  // random
+		"double-4":            float64(math.MaxInt64),
+		"double-6":            doubleMaxVerge,
```
Comment (AlekSi) [inline integration/shareddata/scalars.go:53]:
```suggestion
		"double-5":            doubleMaxVerge,
```

**Review by chilagrow — APPROVED** [review]
Aside from existing comment, looks good to me

File: integration/shareddata/scalars.go (hunk 8 above)
Comment (noisersup) [reply integration/shareddata/scalars.go]: It was already there: https://github.com/FerretDB/FerretDB/pull/3024/files/ac63e6ed60f90c0d87759bead40b480b0b9ee7fb#diff-8307cfe62efa2f64d4ef2d0aff407a8af0986af9f9efe7f98de217e13fe89297R348

File: CONTRIBUTING.md (hunk 1 above)
Comment (chilagrow) [reply CONTRIBUTING.md:260]: I guess no action on this? Perhaps the task of linter if we agree

**Review by chilagrow — APPROVED** [review]

File: CONTRIBUTING.md (hunk 1 above)
Comment (noisersup) [reply CONTRIBUTING.md:260]: > I guess no action on this?

I made all of the constants to use `float64()`

**Review by rumyantseva — APPROVED** [review]
LGTM

**Review by quasilyte — APPROVED** [review]

**Review by AlekSi — APPROVED** [review]
